Sources/MongoSwift/MongoCollection+BulkWrite.swift

```diff
@@ -412,29 +412,54 @@ public struct BulkWriteResult: Codable {
      *   - `MongoError.InternalError` if an unexpected error occurs reading the reply from the server.
      */
     fileprivate init?(reply: BSONDocument, insertedIDs: [Int: BSON]) throws {
-        guard reply.keys.contains(where: { MongocKeys(rawValue: $0) != nil }) else {
-            return nil
-        }
-
-        self.deletedCount = reply[MongocKeys.deletedCount.rawValue]?.toInt() ?? 0
-
-        self.insertedCount = reply[MongocKeys.insertedCount.rawValue]?.toInt() ?? 0
-        self.insertedIDs = insertedIDs
-        self.matchedCount = reply[MongocKeys.matchedCount.rawValue]?.toInt() ?? 0
-        self.modifiedCount = reply[MongocKeys.modifiedCount.rawValue]?.toInt() ?? 0
-        self.upsertedCount = reply[MongocKeys.upsertedCount.rawValue]?.toInt() ?? 0
-
+        var deletedCount: Int?
+        var insertedCount: Int?
+        var matchedCount: Int?
+        var modifiedCount: Int?
+        var upsertedCount: Int?
         var upsertedIDs = [Int: BSON]()
 
-        if let upserted = reply[MongocKeys.upserted.rawValue]?.arrayValue {
-            for upsert in upserted.compactMap({ $0.documentValue }) {
-                guard let index = upsert["index"]?.toInt() else {
-                    throw MongoError.InternalError(message: "Could not cast upserted index to `Int`")
+        var seenKey = true
+        for (k, v) in reply {
```

Comment:
suggest adding a comment somewhere indicating that this method is written this way for perf reasons so we can do it in a single pass, seems maybe easy to forget otherwise

Reply:
done